### PR TITLE
Make some OAuth2 settings optional

### PR DESCRIPTION
**Pull request by MarcialRosales**
## Proposed Changes

Implements features:
- https://github.com/rabbitmq/rabbitmq-server/issues/12235 
- https://github.com/rabbitmq/rabbitmq-server/issues/12237
- https://github.com/rabbitmq/rabbitmq-server/issues/12211

It is accompanied by this docs PR https://github.com/rabbitmq/rabbitmq-website/pull/2056

Tasks:
- [X] Refactor oauth2 plugin so that the core logic uses `oauth_provider` and `resource_server` types rather than asking for each setting to the deprecated config module. Split config module into oauth_provider and resource_server modules.
- [X] Modify oauth2 schema to include `discovery_endpoint_path` and `discovery_endpoint_params`
- [X] Modify management schema to include `oauth_authorization_endpoint_params` and `oauth_token_endpoint_params`
- [X] Modify oauth2_client so that it also accepts params in the `access_token_request` and `discovery_endpoint` to the `oauth_provider` type. (TODO the WSR plugin should be updated to read these extra params and pass them to the `access_token_request`)
- [X] Modify management module that generates authSettings for the ui so that it includes the new schema settings 
- [X] Modify javascript (helper.js) so that it initilizes oidc-client-ts with the new settings
- [X] Verify changes against keycloak + uaa
- [X] Verify changes against auth0
- [X] Verify changes against entra
- [x] Verify changes against okta

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)
- [ ] Build system and/or CI

### CODE REVIEW

File: deps/rabbitmq_auth_backend_oauth2/src/resource_server.erl [hunk 1]
```diff
@@ -0,0 +1,240 @@
+%% This Source Code Form is subject to the terms of the Mozilla Public
+%% License, v. 2.0. If a copy of the MPL was not distributed with this
+%% file, You can obtain one at https://mozilla.org/MPL/2.0/.
+%%
+%% Copyright (c) 2007-2024 Broadcom. All Rights Reserved. The term “Broadcom” refers to Broadcom Inc. and/or its subsidiaries. All rights reserved.
+%%
+
+-module(resource_server).
```
Comment (michaelklishin) [inline deps/rabbitmq_auth_backend_oauth2/src/resource_server.erl:8]: This is a very generic name, we should continue using `rabbit_oauth2_` for prefix.

File: deps/rabbitmq_auth_backend_oauth2/src/rar.erl [hunk 2]
```diff
@@ -0,0 +1,183 @@
+%% This Source Code Form is subject to the terms of the Mozilla Public
+%% License, v. 2.0. If a copy of the MPL was not distributed with this
+%% file, You can obtain one at https://mozilla.org/MPL/2.0/.
+%%
+%% Copyright (c) 2007-2024 Broadcom. All Rights Reserved. The term “Broadcom” refers to Broadcom Inc. and/or its subsidiaries. All rights reserved.
+%%
+
+% Rich Authorization Request
+-module(rar).
```
Comment (michaelklishin) [inline deps/rabbitmq_auth_backend_oauth2/src/rar.erl:9]: This is a very generic and short name, we should continue using `rabbit_oauth2_` for prefix.

File: deps/rabbitmq_auth_backend_oauth2/src/oauth_provider.erl [hunk 3]
```diff
@@ -0,0 +1,197 @@
+%% This Source Code Form is subject to the terms of the Mozilla Public
+%% License, v. 2.0. If a copy of the MPL was not distributed with this
+%% file, You can obtain one at https://mozilla.org/MPL/2.0/.
+%%
+%% Copyright (c) 2007-2024 Broadcom. All Rights Reserved. The term “Broadcom” refers to Broadcom Inc. and/or its subsidiaries. All rights reserved.
+%%
+
+-module(oauth_provider).
```
Comment (michaelklishin) [inline deps/rabbitmq_auth_backend_oauth2/src/oauth_provider.erl:8]: This is a very generic name, we should continue using `rabbit_oauth2_` for prefix.

File: deps/rabbitmq_auth_backend_oauth2/src/keycloak.erl [hunk 4]
```diff
@@ -0,0 +1,41 @@
+%% This Source Code Form is subject to the terms of the Mozilla Public
+%% License, v. 2.0. If a copy of the MPL was not distributed with this
+%% file, You can obtain one at https://mozilla.org/MPL/2.0/.
+%%
+%% Copyright (c) 2007-2024 Broadcom. All Rights Reserved. The term “Broadcom” refers to Broadcom Inc. and/or its subsidiaries. All rights reserved.
+%%
+
+-module(keycloak).
```
Comment (michaelklishin) [inline deps/rabbitmq_auth_backend_oauth2/src/keycloak.erl:8]: This is a very generic name, we should continue using `rabbit_oauth2_` for prefix.

**Review by michaelklishin — CHANGES_REQUESTED** [review]
@MarcialRosales we cannot use very generic modules names such as `rar` and `oauth2`, or `keycloak`.

We should continue using the `rabbit_oauth2_` prefix in this plugin to avoid potential code path conflicts with external libraries, other plugins, and so on.

File: deps/rabbitmq_auth_backend_oauth2/src/resource_server.erl (hunk 1 above)
Comment (MarcialRosales) [reply deps/rabbitmq_auth_backend_oauth2/src/resource_server.erl:8]: I see. I wanted to make it simpler but clearly I went over the line :) . Thanks Michael.

**Comment by michaelklishin** [comment]
@MarcialRosales and I have agreed to move the AWS suite to use Make as part of this PR.

**Comment by michaelklishin** [comment]
Erlang 27 and AWS peer discovery jobs need some work, they are not related to these changes.

**Comment by michaelklishin** [comment]
All suites pass on Erlang 26 and the failures on 27 are well known => merging.

**Review by michaelklishin — APPROVED** [review]

**Pull request merged**